### PR TITLE
Integrate Braintree Venmo

### PR DESCRIPTION
**Pull request by RyanofWoods**
Integrate Braintree's Venmo according to their documentation - https://developer.paypal.com/braintree/docs/guides/venmo

Before, Venmo was only available as a funding source through PayPal, meaning you're forced to use this if you want/need Venmo, and therefore cannot use vaulting. Now it's using Braintree's Venmo integration, stores can benefit from using vaulting Venmo, without affecting the PayPal integration. As it is also separate from the Credit Card and PayPal integrations it also allows for more frontend customization and other benefits such as:
- Venmo to easily be toggled on and off;
- Appearance customization (color and width);
- Venmo payments to be easily plucked in the database as well as through Braintree's transaction control panel.

-----------

Due to this integration, the confirm and complete checkout pages now shows what extension payment type has been used (Credit Card, PayPal, ApplePay and Venmo); for Venmo, it also shows what Venmo account was used.

-----------
Detailed information on integrating Venmo has been added to the README.md

### CODE REVIEW

**Comment by RyanofWoods** [comment]
Hi 
Currently all Spree::Payments that used Venmo in admin show up as "Risky" because no Postal information was given during the transaction(AVS code "I"). I don't see any reference of a specific way to an address in the documentation or the Braintree Venmo tests, so I am not sure what to do about this.

AVS codes:
https://developer.paypal.com/braintree/docs/reference/general/processor-responses/avs-cvv-responses#avs

Venmo integration documentation:
https://developer.paypal.com/braintree/docs/guides/venmo/server-side#creating-transactions

Braintree Venmo transaction tests:
https://github.com/braintree/braintree_ruby/blob/4dca0d777e1c598f10ecdfd9fba8a8c964768ef7/spec/integration/braintree/transaction_spec.rb#L1935

**Comment by DanielePalombo** [comment]
@RyanofWoods did you try to edit the Fraud Management settings?

Screenshot 2022-01-13 at 17 26 59

I'm sure you have already done it.

**Comment by DanielePalombo** [comment]
I have placed an order using the credit card and it has been marked as Risky anyway. It happens because for default the extension provides only the information for the payment without any customer data.
So this is normal even for Venmo.

File: db/migrate/20220104150301_add_venmo_to_braintree_configuration.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddVenmoToBraintreeConfiguration < ActiveRecord::Migration[6.1]
```
Comment (RyanofWoods) [inline db/migrate/20220104150301_add_venmo_to_braintree_configuration.rb:1]: This migration could be lowered to improve backwards compatibility

**Comment by RyanofWoods** [comment]
The AVSResult commit can be dropped if we incorporate https://github.com/solidusio/solidus_paypal_braintree/pull/312 into master beforehand and then rebase.

**Review by DanielePalombo — APPROVED** [review]
LGTM

**Comment by kennyadsl** [comment]
@RyanofWoods can you please iodate the PR description with a brief explanation of the PR content, thanks!

File: README.md [hunk 2]
```diff
@@ -146,6 +147,62 @@ Braintree has some [excellent documentation](https://developers.braintreepayment
 
 For additional information check out [Apple's documentation](https://developer.apple.com/reference/applepayjs/) and [Braintree's documentation](https://developers.braintreepayments.com/guides/apple-pay/client-side/javascript/v3).
 
+## Venmo
+There are two ways to utilize Venmo within this extension:
+- Braintree's native Venmo integration
+  - Offers vaulting of payment sources
+- As a finance option for PayPal
+  - Can only be used checkout flow
```
Comment (kennyadsl) [inline README.md:155]: This point is not super clear to me, maybe it's just missing some words or it's me missing something?

File: README.md [hunk 3]
```diff
@@ -146,6 +147,62 @@ Braintree has some [excellent documentation](https://developers.braintreepayment
...
+  - No vaulting
+
+Note, Venmo is currently available during the payment checkout page (not cart or product pages).
```
Comment (kennyadsl) [inline README.md:158]: This is the same thing stated at line 165, WDYT about removing it?

File: README.md [hunk 4]
```diff
@@ -146,6 +147,62 @@ Braintree has some [excellent documentation](https://developers.braintreepayment
...
+
+### Braintree's Venmo
+#### Note
+- Only available as a financing option on the checkout page; Venmo currently does not support shipping callbacks so it cannot be on the cart page.
+- Currently available to US merchants and buyers and there are also other prequisites.
+  - https://developer.paypal.com/docs/business/checkout/pay-with-venmo/#eligibility
+  - https://developer.paypal.com/braintree/articles/guides/payment-methods/venmo#availability
+
+#### Integration:
+1. Enable Venmo be sure to have it enabled in your [Braintree account](https://developer.paypal.com/braintree/articles/guides/payment-methods/venmo#setup)
```
Comment (kennyadsl) [inline README.md:171]: extra `Enable Venmo` at the beginning of the sentence?

File: lib/solidus_paypal_braintree/engine.rb [hunk 5]
```diff
@@ -20,9 +20,12 @@ class Engine < Rails::Engine
 
     initializer "register_solidus_paypal_braintree_gateway", after: "spree.register.payment_methods" do |app|
       app.config.spree.payment_methods << SolidusPaypalBraintree::Gateway
+      SolidusPaypalBraintree::Gateway.allowed_admin_form_preference_types << :preference_select
       Spree::PermittedAttributes.source_attributes.concat [:nonce, :payment_type, :paypal_funding_source]
     end
 
+    config.assets.precompile += ['solidus_paypal_braintree_manifest.js']
```
Comment (kennyadsl) [inline lib/solidus_paypal_braintree/engine.rb:27]: is this for backend or frontend, maybe we can add below in the already existing `config.assets.precompile` lists?

**Review by kennyadsl — COMMENTED** [review]
Thanks, this looks great. I left some comments, let me know your thougths.

File: lib/solidus_paypal_braintree/engine.rb (hunk 5 above)
Comment (RyanofWoods) [reply lib/solidus_paypal_braintree/engine.rb:27]: The images are only used for frontend currently, so I will move it appropriately, thanks! 👍

**Comment by RyanofWoods** [comment]
Thank you very much for your review @kennyadsl. I made the changes to the precompile and README now. I hope the README is clearer 🙂

**Review by kennyadsl — APPROVED** [review]
Nice work, thanks @RyanofWoods!

**Pull request merged**